File: 25-1C/TDA/TP1/B.hs
import Numeric (showFFloat)

main = do
    s1 <- getLine 
    s2 <- getLine
    let 
        pos1 = count '+' s1 - count '-' s1
        pos2 = count '+' s2 - count '-' s2 
        n = count '?' s2
    putStrLn $ showFFloat Nothing (prob pos1 n pos2) ""

count :: Eq a => a -> [a] -> Int
count e = length.(filter (==e))  

prob :: Int -> Int -> Int -> Float
prob pos1 0 pos2 = if pos1 == pos2 then 1 else 0
prob pos1 n pos2 
    | n < abs (pos1 - pos2) = 0
    | otherwise = (prob pos1 (n-1) (pos2+1) + prob pos1 (n-1) (pos2-1))/2

File: 25-1C/TDA/TP1/B.py
def prob(p1,p2,k):
    if not k: return int(p1==p2)
    if k < abs(p1-p2): return 0
    return (prob(p1,p2+1,k-1)+prob(p1,p2-1,k-1))/2
    
pos1 = input()
pos2 = input()

n = pos2.count("?")
pos1 = pos1.count("+") - pos1.count("-")
pos2 = pos2.count("+") - pos2.count("-")

print(prob(pos1,pos2,n))